hbase-mapreduce/src/main/java/org/apache/hadoop/hbase/mapreduce/RoundRobinTableInputFormat.java

```diff
@@ -0,0 +1,110 @@
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.Map;
+import java.util.HashMap;
+import java.util.List;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.hadoop.hbase.HBaseConfiguration;
+import org.apache.hadoop.hbase.HConstants;
+import org.apache.hadoop.io.RawComparator;
+import org.apache.hadoop.mapreduce.InputFormat;
+import org.apache.hadoop.mapreduce.InputSplit;
+import org.apache.hadoop.mapreduce.JobContext;
+import org.apache.hadoop.mapreduce.JobID;
+import org.apache.hadoop.mapreduce.Mapper;
+import org.apache.hadoop.mapreduce.OutputFormat;
+import org.apache.hadoop.mapreduce.Partitioner;
+import org.apache.hadoop.mapreduce.Reducer;
+import org.apache.hadoop.mapreduce.task.JobContextImpl;
+import org.apache.hadoop.security.Credentials;
+import org.apache.yetus.audience.InterfaceAudience;
+
+/**
+ * Process the lists that come back from {@link TableInputFormat} reordering so splits don't
+ * clump around Servers; try and spread the splits around the cluster.
+ * TIF returns splits in hbase:meta table order. Adjacent or near-adjacent
+ * Regions could be hosted on the same RegionServer. Task placement therefore
+ * could be lumpy with some RegionServers serving lots of inputs where other
+ * servers could be serving few or idle. See the below helpful Flipkart blog post for
+ * a description and from where the base of this code comes from.
+ * @see https://tech.flipkart.com/is-data-locality-always-out-of-the-box-in-hadoop-not-really-2ae9c95163cb
+ */
+@InterfaceAudience.Public
+public class RoundRobinTableInputFormat extends TableInputFormat {
+  @Override
+  public List<InputSplit> getSplits(JobContext context) throws IOException {
+    List<InputSplit> inputSplits = super.getSplits(context);
+    List<InputSplit> roundRobinInputSplits = new ArrayList<>();
+    Map<String, List<InputSplit>> regionServersInputSplits = new HashMap<>();
+    // Prepare a hashmap with each region server as key and list of Input Splits as value
+    if (inputSplits != null && !inputSplits.isEmpty()) {
+      for (InputSplit inputSplit : inputSplits) {
+        if (inputSplit instanceof TableSplit) {
+          String regionServer = ((TableSplit)inputSplit).getRegionLocation();
+          if (regionServer != null && !regionServer.isEmpty()) {
+            regionServersInputSplits.computeIfAbsent(regionServer,
+              k -> new LinkedList<>()).add(inputSplit);
+            continue;
+          }
+        }
+        // If TableSplit or region server not found, add it.
+        roundRobinInputSplits.add(inputSplit);
+      }
+    }
+    while (!regionServersInputSplits.isEmpty()) {
+      Iterator<String> iterator = regionServersInputSplits.keySet().iterator();
+      while (iterator.hasNext()) {
+        String regionServer = iterator.next();
+        List<InputSplit> inputSplitListForRegion = regionServersInputSplits.get(regionServer);
```

Comment:
If it was entrySet iterator, we can avoid this get call.  May be we can change it to be forEach() based simpler loop?  Much lesser code.  or you want keep the code as is in the original blog @saintstack ?

Reply:
There will be an exception when forEach() is used and remove() is called.